### ISSUE TITLE
migration: Improve restore codepath

### ISSUE THREAD

**Issue by sboeuf** [cloud-hypervisor/cloud-hypervisor#4783]
I think the design that was chosen in the first place to perform the snapshot/restore of a virtual machine is slightly flawed on the restore operation.

Having each device implementing the `snapshot()` method from the `Snapshottable` trait is fully appropriate because the idea is to retrieve a state that describes an existing device/component at a given point in time. But having a symmetrical method `restore()` on this same trait doesn't fit the need we have on the restore path. When restoring a VM, we want to create all devices/components directly from the snapshot states. This means using the same constructor `Foobar::new()` for both the standard codepath (no migration) and the restore codepath can't be the right approach. Given this constructor has no information coming from the snapshot, it will create the same device it would when booting a brand new VM, and only later will expect the device to be updated with the right values (through the `restore()` call). Problem with this, the device is created for some time with the wrong values until it gets updated. But that means other components might retrieve the wrong values if they interact with the device too early. Also, this makes the restore codepath very hard to follow, increasing the complexity of maintenance.

The best example to show this approach is inappropriate is by looking at a virtio device. Given such device is split between `VirtioDevice` and `VirtioPciDevice` structures, at some point during the creation of the `VirtioPciDevice` (which happens after the `VirtioDevice` has been created), there's a need to find out the device type and the number of queues this device has. On the restore codepath, that means the device will be created from scratch and therefore the number of queues might not match the amount that was negotiated with the guest when running on the source VM. About the device type, if we have a `Vdpa` device as the `VirtioDevice`, the device type is usually retrieved from the backend, but it's too early to connect with the backend at this point, so we end up not knowing the device type at a time it would have been needed.

If we choose to modify the approach by removing the `restore()` method from the `Snapshottable` trait, we would then provide the snapshot to each constructor so they can directly create the object from the appropriate state. We can also decide to create another constructor called `Foobar::restore()`, but at this point this is an implementation detail. The general idea remains the same by giving every object access to a snapshot so that it can create the object from state.

If we look at the previous example, with this new approach the `VirtioDevice` would be directly created with the right internal values, therefore when `VirtioPciDevice` needs some information from there, it would be retrieving the correct ones.

One interesting reference is the `MemoryManager` because this is the only piece of code that already follows the design I've described. It doesn't implement the `restore()` method, but instead is created from the snapshot state directly.

In terms of how to proceed to move the codebase to this new design, here are the main ideas:
- Provide the `Snapshot` to each device that supports being restored. It doesn't have to do anything with it, but it must have access to it if we want to start implementing an alternative way of constructing the object.
- For every device, implement the alternative constructor and remove the `restore()` implementation.
- Once there's not anymore a single object implementing the `restore()` method, simply remove it from the `Snapshottable` trait.

**Comment by rbradford** [comment]
I think you can use generics to have a consistent name and keep the method on trait.

**Comment by likebreath** [comment]
These are great explanations and examples on why the current restore path is problematic.

> * Provide the Snapshot to each device that supports being restored. It doesn't have to do anything with it, but it must have access to it if we want to start implementing an alternative way of constructing the object.
> * For every device, implement the alternative constructor and remove the restore() implementation.

The most straightforward I can see is we reuse `DeviceManager::create_devices()` and augment it with ability to pass-down `Snapshot` information (such as a `Option`) to each devices being created so that each device can know which constructor to use and have the access to the snapshot information. But I feel that's quite a lot changes and could be messy. Hopefully we can find a clean way to do it.

**Comment by likebreath** [comment]
One thing to mention is that `aarch64` had a requirement of restoring the GIC in the middle of restoring `device_manager` and other devices, and the restoration of GIC happens from the `vm` restoring directly [1]. I am not sure if this would be a hurdle for the proposal, given we are essentially moving to restore devices at creation. @michael2012z @MrXinWang 

[1] https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2811#discussion_r660217130

**Comment by michael2012z** [comment]
> One thing to mention is that `aarch64` had a requirement of restoring the GIC in the middle of restoring `device_manager` and other devices, and the restoration of GIC happens from the `vm` restoring directly [1]. I am not sure if this would be a hurdle for the proposal, given we are essentially moving to restore devices at creation.

The specialty of the [VGIC restoring](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/7a6868979400d47fade9fdec2eb6019d46d4b9ab/vmm/src/vm.rs#L2806) is caused by the requirement that the VGIC has to be created after VCPUs creation. This (the requirement) is also true in the initial boot procedure, see [here](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/7a6868979400d47fade9fdec2eb6019d46d4b9ab/vmm/src/vm.rs#L1217).

In the new design, it would be fine if there is a chance to create and restore the VIGC after restoring the VCPUs.

**Comment by sboeuf** [comment]
Thanks for highlighting this special case. I don't think there will be any issue with it. We will have to provide an `Option<Snapshot>` on the current boot codepath so that the VGIC can be created from the snapshot or from scratch. This whole new design should not modify any of the boot order already established.

**Comment by sboeuf** [comment]
> I think you can use generics to have a consistent name and keep the method on trait.

I don't get how you can use generics in that context? Could you please elaborate?

**Comment by rbradford** [comment]
```rust
trait SnapshotData {
  fn as_any(self) -> Any {}
}

trait Restoreable {
  fn from_snapshot(T: SnapshotData) -> Self {}
}
```

**Comment by sboeuf** [comment]
@rbradford could you please have a look at https://github.com/cloud-hypervisor/cloud-hypervisor/pull/4796/commits/cd8d6fd5f3760a16a90cc4e07712601edaaaf360
I'd like to make sure we have a global agreement before going further (there are lots of places where we can apply the same logic). The patch provides the right snapshot to the device manager, which then passes them down to the virtio-block device. I've implemented the virtio-block update, fully relying on this new design and removing the need for `Block` to implement `restore()` from `Snapshottable` trait.
I couldn't see how to apply your proposal using generics. In general, please let me know if I can simplify/factorize much further the code before jumping on global implementation.

**Comment by sboeuf** [comment]
I've added two items to the list of things that should happen as part of the migration improvement:
>Improve the information saved for each device. Some device might save some useless info while some others might be missing important ones.

It would be nice to improve the quality of the snapshot, removing useless data from being saved, but also this could be an opportunity to enhance the quality of the restore if we could save some important info that were not saved before.

>Flatten the Snapshot hierarchy

I'm anticipating that with the new design, we won't need parent/child hierarchy between `VirtioDevice` and `VirtioPciDevice`, meaning we could flatten the whole Snapshot into a "simple" hash map, rather than a complex hierarchy of devices.

**Comment by sboeuf** [comment]
> > One thing to mention is that `aarch64` had a requirement of restoring the GIC in the middle of restoring `device_manager` and other devices, and the restoration of GIC happens from the `vm` restoring directly [1]. I am not sure if this would be a hurdle for the proposal, given we are essentially moving to restore devices at creation.
> 
> The specialty of the [VGIC restoring](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/7a6868979400d47fade9fdec2eb6019d46d4b9ab/vmm/src/vm.rs#L2806) is caused by the requirement that the VGIC has to be created after VCPUs creation. This (the requirement) is also true in the initial boot procedure, see [here](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/7a6868979400d47fade9fdec2eb6019d46d4b9ab/vmm/src/vm.rs#L1217).
> 
> In the new design, it would be fine if there is a chance to create and restore the VIGC after restoring the VCPUs.

@michael2012z I think I'm hitting some issue related to the vGIC creation happening late with the PR https://github.com/cloud-hypervisor/cloud-hypervisor/pull/4815
Basically, on the new restore path for `MsixConfig`, the code is trying to set irqfd and gsi routes with the following code: https://github.com/cloud-hypervisor/cloud-hypervisor/pull/4815/files#diff-09876698f59b921fd18ac2568ec38cb55354a065a96ffc10486cc9f06fd007fcR106-R116 and it fails with:
```
UpdateInterruptRoute(Custom { kind: Other, error: "Failed registering irq_fd: Failed to register interrupt event: Resource temporarily unavailable (os error 11)"
```
But at this point there's no vGIC, which might be the reason why it fails.

So here is my question, could you describe in details the constraints related to the vGIC creation, as I need to have a proper understanding if I want to come up with a fix/workaround for it :)

**Comment by michael2012z** [comment]
Hi, @sboeuf , here comes some explanation about the vGIC dependency.

2 GIC related types exist in CH:
* [Gic](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/2bbb08b2a4f64ecb41a624309d838f8c261fda2d/devices/src/gic.rs#L40) (implements InterruptController) in `devices/src/gic.rs`, it handles the interrupt source group. I will call it `GicInterruptService` here for easy recognition.
* [Vgic](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/2bbb08b2a4f64ecb41a624309d838f8c261fda2d/hypervisor/src/arch/aarch64/gic.rs#L32) in `hypervisor/src/arch/aarch64/gic.rs`, the hypervisor specific virtual GIC device. I will call it `KvmVGic` in the following text.

AArch64 expects following order for these GIC related instances during the restoration:
1. Restore VCPU - [vm.rs:2704](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/2bbb08b2a4f64ecb41a624309d838f8c261fda2d/vmm/src/vm.rs#L2704)
(Following 2/3/4 are all covered in Vm::restore_vgic_and_enable_interrupt())
2. Create `KvmVgic` (This depends on the creation of VCPU.) - [vm.rs:2254](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/2bbb08b2a4f64ecb41a624309d838f8c261fda2d/vmm/src/vm.rs#L2254)
3. Restore `KvmVgic` - [vm.rs:2283](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/2bbb08b2a4f64ecb41a624309d838f8c261fda2d/vmm/src/vm.rs#L2283)
4. Enable `GicInterruptService` (interrupt source group is enabled in this step) - [vm.rs:2298](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/2bbb08b2a4f64ecb41a624309d838f8c261fda2d/vmm/src/vm.rs#L2298)
5. Restore devices (I suppose the MsixConfig interrupt should be enabled in this step) - [vm.rs:2718](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/2bbb08b2a4f64ecb41a624309d838f8c261fda2d/vmm/src/vm.rs#L2718)

I think the reason of the issue is that the devices restoration came too early for AArch64. GIC things need to be ready before that, and VCPU need to be ready before GIC.

By the way, the same order between VCPU and GIC is also gauranteed in booting process:
1. Create VCPU - [vm.rs:2116](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/2bbb08b2a4f64ecb41a624309d838f8c261fda2d/vmm/src/vm.rs#L2116)
2. Create `KvmVgic` - [vm.rs:1184](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/2bbb08b2a4f64ecb41a624309d838f8c261fda2d/vmm/src/vm.rs#L1184)
3. Enable `GicInterruptService` - [vm.rs:1229](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/2bbb08b2a4f64ecb41a624309d838f8c261fda2d/vmm/src/vm.rs#L1229)

Hope the description help. Please let me know if there is anything I can do.

**Comment by sboeuf** [comment]
Thanks for the detailed explanation @michael2012z 
I'd like to try modifying the sequence so that it's the same for both `aarch64` and `x86_64`. This means the DeviceManager would be created after the CpuManager, but I don't know how much rework/refactoring this is going to need, and if this is even possible.

**Comment by sboeuf** [comment]
@michael2012z I can see the vGIC is created from the `Gic` but that sounds wrong. I mean it'd be best if we could create the vGIC first (of course after the vCPU creation), and then provide the vGIC to the DeviceManager so that it could properly create the `Gic` object. Does that make sense?

**Comment by michael2012z** [comment]
I see. You mean moving the creation of `vGIC` out of `Gic`, and creating the `vGIC` before the `DeviceManager` (and `Gic`),  right?

Yes, I think that works.

We can co-work the refactoring: You adjust the overall sequence as you have designed and make x86_64 work at first, with a new PR. Then I do the aarch64 part. What do you think?

**Comment by sboeuf** [comment]
> I see. You mean moving the creation of `vGIC` out of `Gic`, and creating the `vGIC` before the `DeviceManager` (and `Gic`), right?

Yes exactly.

> 
> Yes, I think that works.
> 
> We can co-work the refactoring: You adjust the overall sequence as you have designed and make x86_64 work at first, with a new PR. Then I do the aarch64 part. What do you think?

That sounds good, I'll keep you posted!

**Comment by sboeuf** [comment]
Hi @michael2012z 
I've created the PR https://github.com/cloud-hypervisor/cloud-hypervisor/pull/4857 (from my [branch](https://github.com/sboeuf/cloud-hypervisor-1/tree/switch_cpu_manager_with_device_manager_creation)) to cover the reordering of the `CpuManager` along with vCPU creation before the `DeviceManager` creation happens. Unfortunately, it breaks the CPU hotplug because of a [deadlock](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/4857#pullrequestreview-1168872228) I can't explain.

I'll be OOO for the next two weeks, but you should be able to work on the move of the vGIC based on what I've done so far. And if you have some spare time, it'd be nice if you could investigate the root cause of the deadlock.

/cc @rbradford

**Comment by michael2012z** [comment]
@sboeuf Thanks. Now I am working on the code base of your branch. 

I added a draft commit (to be verified and finalized) on my working branch https://github.com/michael2012z/cloud-hypervisor/commits/boot-order . Test case `test_snapshot_restore` passed with it.

Next I will test more and investigate the deadlock.

**Comment by michael2012z** [comment]
Got the reason of the deadlock: The lock of the booting VCPUs were never released, see https://github.com/cloud-hypervisor/cloud-hypervisor/blob/1cb1cff8823ebffe87ae5956ed38e74420ca7718/vmm/src/cpu.rs#L970

So a deadlock happened when you try to lock a booting VCPU (which was running) in the loop of `configure_vcpus`. I will come up with a fix with `try_lock()` later.

**Comment by michael2012z** [comment]
Plan for some recent PR works related to GIC:
- Reopen #4875, fixing the deadlock - #4864
- New PR for AArch64 GIC initialization process
- Reopen #4815

**Comment by sboeuf** [comment]
@michael2012z could you move the `Gic` and the associated `KvmGicV3Its` to the new restore design? The idea is to use the function `new()` to handle the restore. (see this as ref: https://github.com/cloud-hypervisor/cloud-hypervisor/commit/ef92e55998a529219a24eb30b13b9cea6c5c1871)

**Comment by michael2012z** [comment]
OK. I will do that tomorrow.

**Comment by michael2012z** [comment]
With my draft code for moving the Gic to the new restore path, the VM failed to restore.  😿 
I am troubleshooting.
The work in progress is here: https://github.com/michael2012z/cloud-hypervisor/tree/restore-gic

**Comment by sboeuf** [comment]
@michael2012z please see my comment https://github.com/michael2012z/cloud-hypervisor/commit/e6cf8efb3dadc74406c9f4f05bc0026e7eccba6c#r91540922

**Comment by sboeuf** [comment]
@rbradford looking at the snapshot code, I was wondering why we ended up with some structures being versionized vs some other structures being serialized? I know we recently moved to versionized structures, but why didn't we move everything to this?

**Comment by rbradford** [comment]
> @rbradford looking at the snapshot code, I was wondering why we ended up with some structures being versionized vs some other structures being serialized? I know we recently moved to versionized structures, but why didn't we move everything to this?

Maybe if you can point me at the particular structure I might be able to tell you if it was intentional or just an oversight.

**Comment by sboeuf** [comment]
@michael2012z @liuw @MrXinWang @likebreath @rbradford 

Hi folks, following the recent redesign of the restore codepath, I've sent the PR https://github.com/cloud-hypervisor/cloud-hypervisor/pull/4970 which takes care of simplifying both `Snapshot` and `SnapshotDataSection` structures. There is some redundancy in carrying identifiers through `Snapshot` and `SnapshotDataSection` when it's only needed to be present as part of the BTreeMap that carries the list of snapshots.
The only drawback of this PR is that it breaks the ability to migrate from previous versions (live upgrade is broken). Are you all okay with that or do you think it's not worth it? Just want to add that aarch64 live upgrade has already been broken last week because we had to move the Gic snapshot/restore to the `DeviceManager`.

**Comment by MrXinWang** [comment]
> Just want to add that aarch64 live upgrade has already been broken last week because we had to move the Gic snapshot/restore to the DeviceManager.

@sboeuf Correct, so I guess it does not make any difference on Arm anyway :) Unless @michael2012z has some specific concerns?

**Comment by sboeuf** [comment]
> > @rbradford looking at the snapshot code, I was wondering why we ended up with some structures being versionized vs some other structures being serialized? I know we recently moved to versionized structures, but why didn't we move everything to this?
> 
> Maybe if you can point me at the particular structure I might be able to tell you if it was intentional or just an oversight.

`Vcpu`, `DeviceManager` and `Gic` invoke `new_from_state()`, and `Vm` invokes `serde_json::to_vec()`. So these are the 4 places where we use non versionized states.

**Comment by rbradford** [comment]
@sboeuf I think improving the quality of the code around snapshot/restore is a great thing do to.

Personally I don't think it's a problem to break cross version migration as our README explicitly says that it is not guaranteed. This issue has been open for some time so I think you only need to wait another 24 hours to hear any dissent.

I did just add this to the plan for v29.0. Ideally this should have been included before cycle began so that everyone was informed of what the goal was. Maybe we need a formal two week window before the the start of a cycle for such extensive changes.

**Comment by likebreath** [comment]
I took a look at PR #4970 and I think its changes are good improvements to make snapshot/restore code paths cleaner. I am fine with merging these changes at the cost of breaking live upgrades.

> `Vcpu`, `DeviceManager` and `Gic` invoke `new_from_state()`, and `Vm` invokes `serde_json::to_vec()`. So these are the 4 places where we use non versionized states.

Maybe it is a good time to make them versionized?

**Comment by rbradford** [comment]
> I took a look at PR #4970 and I think its changes are good improvements to make snapshot/restore code paths cleaner. I am fine with merging these changes at the cost of breaking live upgrades.
> 
> > `Vcpu`, `DeviceManager` and `Gic` invoke `new_from_state()`, and `Vm` invokes `serde_json::to_vec()`. So these are the 4 places where we use non versionized states.
> 
> Maybe it is a good time to make them versionized?

Do those perchance use a Rust style enum in their structs? I think that might not be supported by `#[derive(Versionize)]`

**Comment by rbradford** [comment]
> > > I took a look at PR #4970 and I think its changes are good improvements to make snapshot/restore code paths cleaner. I am fine with merging these changes at the cost of breaking live upgrades.
> > > > `Vcpu`, `DeviceManager` and `Gic` invoke `new_from_state()`, and `Vm` invokes `serde_json::to_vec()`. So these are the 4 places where we use non versionized states.
> > > 
> > > 
> > > Maybe it is a good time to make them versionized?
> > 
> > 
> > Do those perchance use a Rust style enum in their structs? I think that might not be supported by `#[derive(Versionize)]`
> 
> `CpuState` is an enum, `VmSnapshot` contains `ClockData` which is an enum, `DeviceManagerState` contains `DeviceTree` which is a `HashMap<String, DeviceNode>`, and for `Gic` I don't think `Gicv3ItsState` alias `GicState` contains any enum. So except the Gic which might be an easy fix, what are the options for the others?

I think we should try and add support to versionize for Rust enums.

**Comment by sboeuf** [comment]
> > > > I took a look at PR #4970 and I think its changes are good improvements to make snapshot/restore code paths cleaner. I am fine with merging these changes at the cost of breaking live upgrades.
> > > > > `Vcpu`, `DeviceManager` and `Gic` invoke `new_from_state()`, and `Vm` invokes `serde_json::to_vec()`. So these are the 4 places where we use non versionized states.
> > > > 
> > > > 
> > > > Maybe it is a good time to make them versionized?
> > > 
> > > 
> > > Do those perchance use a Rust style enum in their structs? I think that might not be supported by `#[derive(Versionize)]`
> > 
> > 
> > `CpuState` is an enum, `VmSnapshot` contains `ClockData` which is an enum, `DeviceManagerState` contains `DeviceTree` which is a `HashMap<String, DeviceNode>`, and for `Gic` I don't think `Gicv3ItsState` alias `GicState` contains any enum. So except the Gic which might be an easy fix, what are the options for the others?
> 
> I think we should try and add support to versionize for Rust enums.

I had a quick try and I think it already works. So it's a matter of moving the missing pieces to `Versionize`. I'm gonna have a quick try.

**Comment by sboeuf** [comment]
> > > > > I took a look at PR #4970 and I think its changes are good improvements to make snapshot/restore code paths cleaner. I am fine with merging these changes at the cost of breaking live upgrades.
> > > > > > `Vcpu`, `DeviceManager` and `Gic` invoke `new_from_state()`, and `Vm` invokes `serde_json::to_vec()`. So these are the 4 places where we use non versionized states.
> > > > > 
> > > > > 
> > > > > Maybe it is a good time to make them versionized?
> > > > 
> > > > 
> > > > Do those perchance use a Rust style enum in their structs? I think that might not be supported by `#[derive(Versionize)]`
> > > 
> > > 
> > > `CpuState` is an enum, `VmSnapshot` contains `ClockData` which is an enum, `DeviceManagerState` contains `DeviceTree` which is a `HashMap<String, DeviceNode>`, and for `Gic` I don't think `Gicv3ItsState` alias `GicState` contains any enum. So except the Gic which might be an easy fix, what are the options for the others?
> > 
> > 
> > I think we should try and add support to versionize for Rust enums.
> 
> I had a quick try and I think it already works. So it's a matter of moving the missing pieces to `Versionize`. I'm gonna have a quick try.
> 
> EDIT: Ah the blocking piece is that `Versionize` isn't implemented for `HashMap` and for `struct variant` enums.

@rbradford I'm not sure what would be the best approach here, either try to implement both HashMap and all types of enums, or try to versionize only types that can be handled. For instance we could decide to implement our own `Versionize` of the `DeviceTree`.

**Comment by rbradford** [comment]
> > > > > > I took a look at PR #4970 and I think its changes are good improvements to make snapshot/restore code paths cleaner. I am fine with merging these changes at the cost of breaking live upgrades.
> > > > > > > `Vcpu`, `DeviceManager` and `Gic` invoke `new_from_state()`, and `Vm` invokes `serde_json::to_vec()`. So these are the 4 places where we use non versionized states.
> > > > > > 
> > > > > > 
> > > > > > Maybe it is a good time to make them versionized?
> > > > > 
> > > > > 
> > > > > Do those perchance use a Rust style enum in their structs? I think that might not be supported by `#[derive(Versionize)]`
> > > > 
> > > > 
> > > > `CpuState` is an enum, `VmSnapshot` contains `ClockData` which is an enum, `DeviceManagerState` contains `DeviceTree` which is a `HashMap<String, DeviceNode>`, and for `Gic` I don't think `Gicv3ItsState` alias `GicState` contains any enum. So except the Gic which might be an easy fix, what are the options for the others?
> > > 
> > > 
> > > I think we should try and add support to versionize for Rust enums.
> > 
> > 
> > I had a quick try and I think it already works. So it's a matter of moving the missing pieces to `Versionize`. I'm gonna have a quick try.
> > EDIT: Ah the blocking piece is that `Versionize` isn't implemented for `HashMap` and for `struct variant` enums.
> 
> @rbradford I'm not sure what would be the best approach here, either try to implement both HashMap and all types of enums, or try to versionize only types that can be handled. For instance we could decide to implement our own `Versionize` of the `DeviceTree`.

I don't know. What do you think? The latter might be easier.

**Comment by sboeuf** [comment]
> > > > I think we should try and add support to versionize for Rust enums.
> > > 
> > > 
> > > I had a quick try and I think it already works. So it's a matter of moving the missing pieces to `Versionize`. I'm gonna have a quick try.
> > > EDIT: Ah the blocking piece is that `Versionize` isn't implemented for `HashMap` and for `struct variant` enums.
> > 
> > 
> > @rbradford I'm not sure what would be the best approach here, either try to implement both HashMap and all types of enums, or try to versionize only types that can be handled. For instance we could decide to implement our own `Versionize` of the `DeviceTree`.
> 
> I don't know. What do you think? The latter might be easier.

I tend to agree.

**Comment by rbradford** [comment]
Our live migration/snapshot & restore code paths have had a lot of improvements recently. Thanks @phip1611! So I think this issue no longer applies or can be considered stale
